operator/v1alpha1/types_image_content_source_policy.go

```diff
@@ -0,0 +1,46 @@
+package v1alpha1
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ImageContentSource holds cluster-wide information about how to handle registry mirror rules.  The canonical name is `cluster`
+type ImageContentSource struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object's metadata.
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec holds user settable values for configuration
+	// +required
+	Spec ImageContentSourceSpec `json:"spec"`
+}
+
+type ImageContentSourceSpec struct {
+	// MirrorReposForDigests holds the registry/repo mirroring rules that are written to
+	// /etc/containers/registries.conf to be used by the runtime to determine which
+	// repo to fall back on, if the original repo cannot be accessed.
+	// We only support pulling by digests for mirrors.
+	// Will also be used in cases such as offline installs/upgrades.
+	// +optional
+	MirrorReposForDigests MirrorReposForDigests `json:"mirrorReposForDigests"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+type ImageContentSourceList struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object's metadata.
+	metav1.ListMeta `json:"metadata"`
+	Items           []ImageContentSource `json:"items"`
+}
+
+// MirrorReposForDigests holds cluster-wide information about how to handle mirros in the registries config.
+// Note: the mirrors only work when pulling the imageby its digest
+type MirrorReposForDigests struct {
+	// Sources are registries and/or repositories depending on the container runtime
+	// that are mirrors of each other.
+	// +optional
+	Sources []string `json:"sources,omitempty"`
```

Comment:
This needs to be a list of lists, the parent (in ImageContentSource.Spec) should be an array (instead of a nested struct).

Reply:
Fixed, made `repositoryDigestMirrors` a list as well.

Reply:
> This needs to be a list of lists, the parent (in ImageContentSource.Spec) should be an array (instead of a nested struct).

…
> There should be no canonical name for this - clients may create multiple of these, it is not a singleton. We want individual components to be able to register their policies and manage them independently

(I know very little about k8s API conventions): Isn’t a list of objects that contains a list of entries unnecessarily too much to iterate over? If a client can create multiple instances of `ImageContentSourcePolicy`, why does it need `ImageContentSourcePolicySpec.RepositoryDigestMirrors` to be an array?

Is there a semantic difference between a set of specs vs a single spec with an array?

Reply:
We don't force denormalization of ImageContentSourcePolicy-like objects into multiple objects because that puts an extra burden on users/integrators.  The atomic unit of transaction is a single object in etcd, and so someone who wishes to add / remove a strict policy will naturally do so via GET/PUT operations on a single resource, where they can ensure the policy matches the desired state.